File: app/src/main/java/me/liuyun/bjutlgn/db/UserManager.java
package me.liuyun.bjutlgn.db;

import android.content.Context;
import android.util.Log;

import java.util.List;

import me.liuyun.bjutlgn.entity.User;

public class UserManager {
    private static final String TAG = "UserManager";

    private DBHelper dbHelper;

    public UserManager(Context context) {
        this.dbHelper = new DBHelper(context);
    }


    public boolean insertUser(String account, String password, int pack) {
        try {
            int position = Integer.parseInt(dbHelper.getUserDao().queryRaw(dbHelper.getUserDao().queryBuilder().selectRaw("MAX(position)").prepareStatementString()).getFirstResult()[0]);
            dbHelper.getUserDao().createOrUpdate(new User(0, account, password, pack, position + 1));
            return true;
        } catch (Exception e) {
            Log.e(TAG, "insertUser", e);
        }
        return false;
    }

    public boolean updateUser(User user) {
        try {
            dbHelper.getUserDao().update(user);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "updateUser", e);
        }
        return false;
    }

    public boolean deleteUser(int id) {
        try {
            dbHelper.getUserDao().deleteById(id);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "deleteUser", e);
        }
        return false;
    }

    public List<User> getAllUsers() {
        try {
            return dbHelper.getUserDao().queryBuilder().orderBy("position", true).query();
        } catch (Exception e) {
            Log.e(TAG, "getAllUsers", e);
        }
        return null;
    }

}

File: app/src/main/kotlin/me/liuyun/bjutlgn/db/UserManager.kt
package me.liuyun.bjutlgn.db

import android.content.Context
import android.util.Log
import com.j256.ormlite.dao.Dao

import me.liuyun.bjutlgn.entity.User

class UserManager(context: Context) {

    private val dbHelper: DBHelper = DBHelper(context)

    fun transaction(body: (Dao<User, Int>) -> Int): Int {
        try {
            return body(dbHelper.userDao)
        } catch (e: Exception) {
            Log.e(TAG, body.toString(), e)
        }
        return 0
    }

    fun <T> transaction(body: (Dao<User, Int>) -> MutableList<T>): MutableList<T> {
        try {
            return body(dbHelper.userDao)
        } catch (e: Exception) {
            Log.e(TAG, body.toString(), e)
        }
        return mutableListOf()
    }


    fun insertUser(user: User): Int {
        return transaction {
            val position = (it.queryRaw(dbHelper.userDao.queryBuilder().selectRaw("MAX(position)").prepareStatementString()).firstResult[0] ?: "-1").toInt()
            user.position = position + 1
            dbHelper.userDao.createOrUpdate(user).numLinesChanged
        }
    }

    fun updateUser(user: User): Int {
        return transaction { it.update(user) }
    }

    fun deleteUser(id: Int): Int {
        return transaction { it.deleteById(id) }
    }

    fun allUsers(): MutableList<User> {
        return transaction<User> { it.queryBuilder().orderBy("position", true).query() ?: mutableListOf() }
    }

    val TAG: String = this.javaClass.name

}
